libraries/core_libs/consensus/src/vote_manager/vote_manager.cpp

```diff
@@ -471,6 +490,182 @@ uint64_t VoteManager::roundDeterminedFromVotes(size_t two_t_plus_one) {
+  std::unique_lock lock(reward_votes_mutex_);
+  if (reward_votes_.second.contains(vote_hash)) {
+    return false;
+  }
+
+  const auto voted_block_hash = vote->getBlockHash();
+  const auto pbft_chain_last_block_hash = pbft_chain_->getLastPbftBlockHash();
+  if (voted_block_hash != pbft_chain_last_block_hash) {
+    LOG(log_dg_) << "Drop reward vote " << vote_hash << ". PBFT chain last block is " << pbft_chain_last_block_hash
+                 << ", but vote for PBFT block is " << voted_block_hash;
+    return false;
+  }
+
+  if (!reward_votes_.first) {
+    reward_votes_.first = voted_block_hash;
+  } else if (reward_votes_.first != pbft_chain_last_block_hash) {
+    // Clear missing reward votes
+    LOG(log_dg_) << "Remove reward votes: \n";
+    for (const auto& v : reward_votes_.second) {
+      LOG(log_dg_) << v.first;
+    }
+    reward_votes_ = {};
+    reward_votes_.first = voted_block_hash;
+  }
+  assert(voted_block_hash == reward_votes_.first);
+
+  reward_votes_.second.insert({vote_hash, vote});
+  LOG(log_nf_) << "add reward vote " << vote_hash;
+
+  return true;
+}
+
+bool VoteManager::verifyRewardVote(std::shared_ptr<Vote>& vote) {
+  const auto voted_block_hash = vote->getBlockHash();
+  const auto pbft_chain_last_block_hash = pbft_chain_->getLastPbftBlockHash();
+  if (voted_block_hash != pbft_chain_last_block_hash) {
+    LOG(log_er_) << "Drop reward vote " << vote->getHash() << ". PBFT chain last block is "
+                 << pbft_chain_last_block_hash << ", but vote for PBFT block is " << voted_block_hash;
+    return false;
+  }
+
+  if (vote->getType() != cert_vote_type) {
+    LOG(log_er_) << "Reward vote type is not cert vote type " << *vote;
+    return false;
+  }
+
+  if (vote->getStep() != 3) {
+    LOG(log_er_) << "Reward vote step is not 3 " << *vote;
+    return false;
+  }
+
+  uint64_t dpos_period = pbft_chain_->getPbftChainSize() - 1;  // reward period - 1
+  const auto& voter_account_addr = vote->getVoterAddr();
+  uint64_t voter_dpos_votes_count;
+  try {
+    voter_dpos_votes_count = final_chain_->dpos_eligible_vote_count(dpos_period, voter_account_addr);
+  } catch (state_api::ErrFutureBlock& c) {
+    LOG(log_er_) << c.what() << ". Voter account " << voter_account_addr << " in DPOS period<" << dpos_period
+                 << "> is too far for ahead of DPOS. Have executed chain size " << final_chain_->last_block_number();
+    return false;
+  }
+
+  if (!voter_dpos_votes_count) {
+    LOG(log_er_) << "Invalid reward vote: voter was not eligible to vote " << *vote;
+    return false;
+  }
+
+  const uint64_t reward_period_dpos_total_votes_count = final_chain_->dpos_eligible_total_vote_count(dpos_period);
+  const size_t reward_period_pbft_sortition_threshold =
+      db_->getPbftSortitionThreshold(dpos_period + 1);  // reward period
+  if (!reward_period_pbft_sortition_threshold) {
+    LOG(log_er_) << "Cannot get PBFT sortition threshold for period " << dpos_period;
```

Comment:
So @gaoqi8649 you mentioned this https://github.com/Taraxa-project/taraxa-node/pull/1670/files#r858131884 but you still compare it with final_chain, so I do not understand meaning of PBFT threshold

Reply:
I think you are right. Fixed in https://github.com/Taraxa-project/taraxa-node/pull/1710
Let me test it first

Reply:
Fixed